pkg/admissioncontroller/webhooks/admission/namespacedeletion/admission.go

```diff
@@ -36,109 +36,107 @@ import (
 )
 
 const (
-	namespaceValidatorName = "namespace_validator"
+	// PluginName is the name of this admission plugin.
+	PluginName = "namespace_validator"
+	// WebhookPath is the HTTP handler path for this admission webhook handler.
+	WebhookPath = "/webhooks/validate-namespace-deletion"
 )
 
 var namespaceGVK = metav1.GroupVersionKind{Group: "", Kind: "Namespace", Version: "v1"}
 
-type namespaceDeletionHandler struct {
-	cacheReader client.Reader
-	apiReader   client.Reader
-
-	logger logr.Logger
-}
-
-// NewValidateNamespaceDeletionHandler creates a new handler for validating namespace deletions.
-func NewValidateNamespaceDeletionHandler(ctx context.Context, cache cache.Cache) (admission.Handler, error) {
+// New creates a new handler for validating namespace deletions.
+func New(ctx context.Context, logger logr.Logger, cache cache.Cache) (admission.Handler, error) {
 	// Initialize caches here to ensure the readyz informer check will only succeed once informers required for this
 	// handler have synced so that http requests can be served quicker with pre-syncronized caches.
-	_, err := cache.GetInformer(ctx, &gardencorev1beta1.Project{})
-	if err != nil {
+	if _, err := cache.GetInformer(ctx, &gardencorev1beta1.Project{}); err != nil {
 		return nil, err
 	}
 
-	return &namespaceDeletionHandler{
+	return &plugin{
 		cacheReader: cache,
+		logger:      logger,
 	}, nil
 }
 
-// InjectLogger injects a logger into the handler.
-func (h *namespaceDeletionHandler) InjectLogger(l logr.Logger) error {
-	h.logger = l.WithName(namespaceValidatorName)
-	return nil
+type plugin struct {
+	cacheReader client.Reader
+	apiReader   client.Reader
+	logger      logr.Logger
 }
 
+var _ admission.Handler = &plugin{}
+
 // InjectAPIReader injects a reader into the handler.
-func (h *namespaceDeletionHandler) InjectAPIReader(reader client.Reader) error {
-	h.apiReader = reader
+func (p *plugin) InjectAPIReader(reader client.Reader) error {
+	p.apiReader = reader
 	return nil
 }
 
 // Handle implements the webhook handler for namespace deletion validation.
-func (h *namespaceDeletionHandler) Handle(ctx context.Context, request admission.Request) admission.Response {
+func (p *plugin) Handle(ctx context.Context, request admission.Request) admission.Response {
 	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
 	defer cancel()
 
 	// If the request does not indicate the correct operation (DELETE) we allow the review without further doing.
 	if request.Operation != admissionv1.Delete {
-		return admissionAllowed("operation is not DELETE")
+		return admission.Allowed("operation is not DELETE")
```

Comment:
FYI: I chose to have a custom `admissionAllowed` func instead of using the upstream funcs, as they put the string into the `Reason` field, which is not 100% correct if we use such description like this one. According to the [API docs](https://github.com/gardener/gardener/blob/03794c4fab012f16625540cd2989ab8f328be93d/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L644-L649), `Reason` is supposed to be a machine-readable description, while `Message` is the human-readable counterpart.
The webhook plugins and kubectl also build on that and transport error messages / reason accordingly, which results in weird error outputs for denied webhook responses.

Reply:
Hm, OK, I saw this as well but was wondering why the controller-runtime doesn't support this then? Shall we switch back to our "custom" function? We don't use this in the GSAC, however, GSAC runs in the seed while GAC is in the garden cluster and its responses could be visible to end-users.

Reply:
Yeah, I was asking myself the same question. But I haven't found the time to aks/discuss/contribute upstream yet.
I would probably keep the custom funcs for now in GAC because of the enduser visibility, however I'm not insisting on it.

Reply:
OK, makes sense, will do.

Reply:
Done in af6adcd